woocommerce/woocommerce-ios
# Filter Products: Yosemite & Networking layers

- [x] ⚠️ update PR base to `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2204 is merged ⚠️ 

Update products tab predicate based on filters & Sync products by each filter via Yosemite for #2037 

This PR implements the changes in Yosemite & Networking layers to filter the Products tab results and also sync products based on the filters.

## Changes

- Networking layer:
  - Added the filter params to `ProductsRemote.loadAllProducts`
- Yosemite layer:
  - Added the filter params to `ProductAction.synchronizeProducts`
  - In `ProductStore`, passed the corresponding filter params to `ProductsRemote`
  - Added a `NSPredicate` helper to create a predicate with the new filters + site ID, and a `ResultsController` helper to update the predicate
  - Tests:
    - Added `productStatus` and `productType` params to `MockProduct`
    - Added a `MockupNetwork` extension helper to get the path components
    - Created `ProductStore_FilterProductsTests` to verify the params are as expected for the filters
    - Created `ResultsController_FilterProductTests` to verify the predicate is filtering the products correctly
- App layer:
  - When the `filters` changes, updated the results controller predicate and also resync the data with the filter values
  - Initialized the results controller with the filter values as well

## Testing

- Go to the Products tab
- Tap on the "Filter" CTA in the toolbar below the WIP banner --> should see the Filter Products screen **without "Clear All" CTA in the navigation bar**
- Tap on "Stock status" row --> it should navigate to the stock status list selector screen
- Select a different stock status (e.g. "Out of stock") filter and tap "Show Products" at the bottom --> it should go back to the Products tab with only products that match the stock status
- Tap on the "Filter" CTA again --> should see the Filter Products screen **with "Clear All" CTA in the navigation bar**, the navigation bar title should say "Filters (1)", and the "Stock status" row should say the selected stock status
- Repeat the same steps for "Product status" and "Product type" rows and notice the navigation bar title should update to the number of active filters, and the Products tab shows the products that match the filters
- Tap "Clear all" CTA in the navigation bar --> all filters should be cleared to "Any"
- Tap "Show Products" at the bottom --> it should go back to the Products tab with all the products

## Example screencast

![simulator](https://user-images.githubusercontent.com/1945542/80555374-b3d6fc00-8a02-11ea-9f67-949aa2f9dadb.gif)



Update release notes:
